src/components/rating/rating.tsx

```diff
@@ -238,27 +254,115 @@ export class Rating
     this.setFocus();
   }
 
-  private updateValue(value: number) {
-    this.value = value;
-    this.calciteRatingChange.emit({ value });
-  }
+  private handleRatingPointerOver = () => {
+    this.isKeyboardInteraction = false;
+  };
 
-  private onKeyboardPressed = (event: KeyboardEvent): void => {
-    if (!this.required && isActivationKey(event.key)) {
-      event.preventDefault();
-      this.updateValue(0);
-    }
+  private handleRatingPointerOut = () => {
+    this.isKeyboardInteraction = true;
+    this.hoverValue = null;
+    this.focusValue = null;
+    this.hasFocus = false;
   };
 
-  private onFocusChange = (selectedRatingValue: number): void => {
+  private handleRatingFocusIn = (): void => {
+    let selectedInput = this.inputFocusRef;
+
+    if (!selectedInput) {
```

Comment:
Instead of traversing the shadowRoot, my suggestion would be to have two private arrays where we could store references to the input and label elements. Then we don't need to have this conditional `(!selectedInput)` logic that has to traverse the DOM.

```
starInputs: HTMLCalciteInputElement = []
starLabels: HTMLCalciteLabelInputElement = []
```

On render, use the `ref` attribute to store them in the array. 

```
ref={(el) =>
              starInputs[i] = el as HTMLInputElement)
            }
```

Then, all references to inputs and labels are available at all times and the DOM doesn't need to be traversed.

Reply:
@benelan @eriklharper what do you think? If you're ok with querying the shadowRoot here then we can go with this.

I was just thinking that we are rendering the elements so we should just store references to them.

Reply:
> Instead of traversing the shadowRoot, my suggestion would be to have two private arrays 

I've added starInputs array to the rendering but I don't see a use-case for starLabels